src/spyglass/behavior/moseq.py

```diff
@@ -0,0 +1,360 @@
+        )
+
+        config = lambda: kpms.load_config(project_dir)
+
+        # fetch the data and format it for moseq
+        coordinates, confidences = PoseGroup().fetch_pose_datasets(
+            key, format_for_moseq=True
+        )
+        data, metadata = kpms.format_data(coordinates, confidences, **config())
+
+        # either initialize a new model or load an existing one
+        model_name = self._make_model_name(key)
+        initial_model_key = model_params.get("initial_model", None)
+        if initial_model_key is None:
+            # fit pca of data
+            pca = kpms.fit_pca(**data, **config())
+            kpms.save_pca(pca, project_dir)
+
+            # create the model
+            model = kpms.init_model(data, pca=pca, **config())
+            # run the autoregressive fit on the model
+            num_ar_iters = model_params["num_ar_iters"]
+            model, model_name = kpms.fit_model(
+                model,
+                data,
+                metadata,
+                project_dir,
+                ar_only=True,
+                num_iters=num_ar_iters,
+                model_name=model_name + "_ar",
+            )
+            epochs_trained = num_ar_iters
+            # load model checkpoint
+            # model, data, metadata, current_iter = kpms.load_checkpoint(
+            #     project_dir, model_name, iteration=num_ar_iters
+            # )
+
+        else:
+            # begin training from an existing model
+            query = MoseqModel & initial_model_key
+            if not query:
+                raise ValueError(
+                    f"Initial model: {initial_model_key} not found"
+                )
+            model = query.fetch_model()
+            epochs_trained = query.fetch1("epochs_trained")
+
+        # update the hyperparameters
+        kappa = model_params["kappa"]
+        model = kpms.update_hypparams(model, kappa=kappa)
+        # run fitting on the complete model
+        num_epochs = model_params["num_epochs"]
+        model = kpms.fit_model(
+            model,
+            data,
+            metadata,
+            project_dir,
+            model_name,
+            ar_only=False,
+            start_iter=epochs_trained,
+            num_iters=epochs_trained + num_epochs,
+        )[0]
+        # reindex syllables by frequency
+        kpms.reindex_syllables_in_checkpoint(project_dir, model_name)
+        self.insert1(
+            {
+                **key,
+                "project_dir": project_dir,
+                "epochs_trained": num_epochs + epochs_trained,
+                "model_name": model_name,
+            }
+        )
+
+    def _make_model_name(self, key: dict = None):
+        # make a unique model name based on the key
+        if key is None:
+            key = {}
+        key = (MoseqModelSelection & key).fetch1("KEY")
+        key_string = "_".join(key.values())
+        return hashlib.sha1(key_string.encode("utf-8")).hexdigest()[:10]
```

Comment:
I'd prefer to use datajoint's builtin hasher in these cases, rather than make a new one to handle unhashable types like dict

`dj.hash.key_hash`